From 7085faa6afd6a54493d6e35347eec10603f76856 Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Tue, 30 Aug 2022 10:22:54 +0300
Subject: [PATCH] arm64: dts: imx8dxl-v2x: enable sja1110a ethernet switch

Describe rthe SJA1110A Ethernet Switch connected to secondary SoC NIC.
3 ports are enabled:
- Port 1: 100Base-TX to external connector
- Port 2: RGMII to SoC NIC
- Port 10: T1 to external connector

Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 arch/arm64/boot/dts/freescale/imx8dxl-v2x.dts | 273 +++++++++++++++++-
 1 file changed, 261 insertions(+), 12 deletions(-)

diff --git a/arch/arm64/boot/dts/freescale/imx8dxl-v2x.dts b/arch/arm64/boot/dts/freescale/imx8dxl-v2x.dts
index c7ebfaeb41a7..dfe3809a3039 100644
--- a/arch/arm64/boot/dts/freescale/imx8dxl-v2x.dts
+++ b/arch/arm64/boot/dts/freescale/imx8dxl-v2x.dts
@@ -56,6 +56,18 @@ regulator_flexcan2_standby: regulator-flexcan2-standby {
 	};
 };
 
+&eqos {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_eqos>;
+	phy-connection-type = "rgmii-id";
+	status = "okay";
+
+	fixed-link {
+		speed = <1000>;
+		full-duplex;
+	};
+};
+
 &flexcan1 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_flexcan1>;
@@ -81,6 +93,31 @@ can-transceiver {
 };
 
 &iomuxc {
+		pinctrl_eqos: eqosgrp {
+			fsl,pins = <
+				/* MDIO to Switch */
+				/* enet0 mdio pads supplied with 3.3v */
+				/* IMX8DXL_COMP_CTL_GPIO_1V8_3V3_GPIOCT */
+				IMX8DXL_ENET0_MDC_CONN_EQOS_MDC				0x06000020
+				IMX8DXL_ENET0_MDIO_CONN_EQOS_MDIO			0x06000020
+				/* RGMII to Switch */
+				/* ENET1 pad supplied with 3.3V */
+				/*IMX8DXL_COMP_CTL_GPIO_1V8_3V3_GPIORHB		0x000514a0*/
+				IMX8DXL_ENET1_RGMII_TX_CTL_CONN_EQOS_RGMII_TX_CTL	0x06000020
+				IMX8DXL_ENET1_RGMII_TXC_CONN_EQOS_RGMII_TXC		0x06000020
+				IMX8DXL_ENET1_RGMII_TXD0_CONN_EQOS_RGMII_TXD0		0x06000020
+				IMX8DXL_ENET1_RGMII_TXD1_CONN_EQOS_RGMII_TXD1		0x06000020
+				IMX8DXL_ENET1_RGMII_TXD2_CONN_EQOS_RGMII_TXD2		0x06000020
+				IMX8DXL_ENET1_RGMII_TXD3_CONN_EQOS_RGMII_TXD3		0x06000020
+				IMX8DXL_ENET1_RGMII_RXC_CONN_EQOS_RGMII_RXC		0x06000020
+				IMX8DXL_ENET1_RGMII_RX_CTL_CONN_EQOS_RGMII_RX_CTL	0x06000020
+				IMX8DXL_ENET1_RGMII_RXD0_CONN_EQOS_RGMII_RXD0		0x06000020
+				IMX8DXL_ENET1_RGMII_RXD1_CONN_EQOS_RGMII_RXD1		0x06000020
+				IMX8DXL_ENET1_RGMII_RXD2_CONN_EQOS_RGMII_RXD2		0x06000020
+				IMX8DXL_ENET1_RGMII_RXD3_CONN_EQOS_RGMII_RXD3		0x06000020
+			>;
+		};
+
 		pinctrl_flexcan1: flexcan1grp {
 			fsl,pins = <
 				IMX8DXL_FLEXCAN0_TX_ADMA_FLEXCAN0_TX	0x00000021
@@ -98,7 +135,7 @@ IMX8DXL_FLEXCAN1_RX_ADMA_FLEXCAN1_RX	0x00000021
 		pinctrl_i2c2_gpio: i2c2gpiogrp {
 			fsl,pins = <
 				/* gpio-expander reset: io without pull-up */
-				IMX8DXL_SCU_GPIO0_00_LSIO_GPIO2_IO03		0x0000061 /* some reason this fails in pinctrl driver */
+				IMX8DXL_SCU_GPIO0_00_LSIO_GPIO2_IO03		0x0000061
 
 				/* gpio-expander interrupt: io without pull-up */
 				/* not usable, because this pin does not support lsio gpio mode :( */
@@ -122,17 +159,6 @@ IMX8DXL_UART1_RTS_B_ADMA_UART1_RTS_B	0x06000020
 			>;
 		};
 
-		pinctrl_strange: strangegrp {
-			fsl,pins = <
-				/* RST_N SJA1110AEL */
-				IMX8DXL_USB_SS3_TC0_LSIO_GPIO4_IO03		0x00000021
-
-				/* RST_CORE_N SJA1110AEL */
-				/*IMX8DXL_USB_SS3_TC1_LSIO_GPIO4_IO04			0x00000021*/
-				IMX8DXL_USB_SS3_TC1_CONN_USB_OTG2_PWR	0x00000021
-			>;
-		};
-
 		pinctrl_lte: ltegrp {
 			fsl,pins = <
 				/* modem RESET_N: io open drain drive 2mA */
@@ -150,6 +176,28 @@ IMX8DXL_SPI3_SDO_LSIO_GPIO0_IO14	0x0000061
 			>;
 		};
 
+		pinctrl_lpspi0: lpspi0grp {
+			fsl,pins = <
+				IMX8DXL_SPI0_SCK_ADMA_SPI0_SCK					0x600004c
+				IMX8DXL_SPI0_SDO_ADMA_SPI0_SDO					0x600004c
+				IMX8DXL_SPI0_SDI_ADMA_SPI0_SDI					0x600004c
+				/* IMX8DXL_SPI0_CS0_ADMA_SPI0_CS0					0x600004c */
+				/* IMX8DXL_SPI0_CS1_ADMA_SPI0_CS1					0x600004c */
+				IMX8DXL_SPI0_CS0_LSIO_GPIO1_IO08				0x0000021
+				IMX8DXL_SPI0_CS1_LSIO_GPIO1_IO07				0x0000021
+			>;
+		};
+
+		pinctrl_lpspi0_switch: lpspi0switchgrp {
+			fsl,pins = <
+				/* SW_RSTn: io without pull-up */
+				IMX8DXL_USB_SS3_TC0_LSIO_GPIO4_IO03				0x0000021
+
+				/* SW_CORE_RSTn: io without pull-up */
+				IMX8DXL_USB_SS3_TC1_LSIO_GPIO4_IO04				0x0000021
+			>;
+		};
+
 		pinctrl_usdhc3: usdhc3grp {
 			fsl,pins = <
 				IMX8DXL_ENET0_RGMII_TXC_CONN_USDHC2_CLK			0x06000040
@@ -238,6 +286,207 @@ &i2c3 {
 	/* nothing connected */
 };
 
+&lpspi0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_lpspi0>, <&pinctrl_lpspi0_switch>;
+	status = "okay";
+	cs-gpios = <&lsio_gpio1 8 GPIO_ACTIVE_LOW>, <&lsio_gpio1 7 GPIO_ACTIVE_LOW>;
+/*
+	spidev@0 {
+		compatible = "rohm,dh2228fv";
+		reg = <0>;
+		spi-max-frequency = <4000000>;
+	};
+
+	spidev@1 {
+		compatible = "rohm,dh2228fv";
+		reg = <1>;
+		spi-max-frequency = <4000000>;
+*/
+
+	ethernet-switch@0 {
+		compatible = "nxp,sja1110a";
+		reg = <0>;
+		//spi-max-frequency = <25000000>;
+		spi-max-frequency = <4000000>;
+		reset-gpios = <&lsio_gpio4 3 GPIO_ACTIVE_LOW>, <&lsio_gpio4 4 GPIO_ACTIVE_LOW>;
+
+		ethernet-ports {
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			/* to Switch uC */
+			port@0 {
+				reg = <0x0>;
+				status = "disabled";
+				label = "uc";
+			};
+
+			/* to RJ45 connector on V2X Adapter, 100BASE-TX */
+			port@1 {
+				reg = <0x1>;
+				status = "okay";
+				label = "lan1";
+				phy-mode = "internal";
+				phy-handle = <&switch_port1_base_tx_phy>;
+			};
+
+			/* to CPU */
+			port@2 {
+				reg = <0x2>;
+				status = "okay";
+				label = "cpu";
+				phy-mode = "rgmii-id";
+				rx-internal-delay-ps = <2000>;
+				tx-internal-delay-ps = <2000>;
+				ethernet = <&eqos>;
+
+				fixed-link {
+					speed = <1000>;
+					full-duplex;
+				};
+			};
+
+			port@3 {
+				reg = <0x3>;
+				status = "disabled";
+				label = "lan3";
+			};
+
+			port@4 {
+				reg = <0x4>;
+				status = "disabled";
+				label = "lan4";
+			};
+
+			port@5 {
+				reg = <0x5>;
+				status = "disabled";
+				label = "trx1";
+				phy-mode = "internal";
+				phy-handle = <&switch_port5_base_t1_phy>;
+			};
+
+			port@6 {
+				reg = <0x6>;
+				status = "disabled";
+				label = "trx2";
+				phy-mode = "internal";
+				phy-handle = <&switch_port6_base_t1_phy>;
+			};
+
+			port@7 {
+				reg = <0x7>;
+				status = "disabled";
+				label = "trx3";
+				phy-mode = "internal";
+				phy-handle = <&switch_port7_base_t1_phy>;
+			};
+
+			port@8 {
+				reg = <0x8>;
+				status = "disabled";
+				label = "trx4";
+				phy-mode = "internal";
+				phy-handle = <&switch_port8_base_t1_phy>;
+			};
+
+			port@9 {
+				reg = <0x9>;
+				status = "disabled";
+				label = "trx5";
+				phy-mode = "internal";
+				phy-handle = <&switch_port9_base_t1_phy>;
+			};
+
+			port@a {
+				reg = <0xa>;
+				status = "okay";
+				label = "trx6";
+				phy-mode = "internal";
+				phy-handle = <&switch_port10_base_t1_phy>;
+			};
+		};
+
+		mdios {
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			mdio@0 {
+				compatible = "nxp,sja1110-base-t1-mdio";
+				#address-cells = <1>;
+				#size-cells = <0>;
+				reg = <0>;
+				status = "okay";
+
+				switch_port5_base_t1_phy: ethernet-phy@1 {
+					compatible = "ethernet-phy-ieee802.3-c45";
+					reg = <0x1>;
+					status = "disabled";
+				};
+
+				switch_port6_base_t1_phy: ethernet-phy@2 {
+					compatible = "ethernet-phy-ieee802.3-c45";
+					reg = <0x2>;
+					status = "disabled";
+				};
+
+				switch_port7_base_t1_phy: ethernet-phy@3 {
+					compatible = "ethernet-phy-ieee802.3-c45";
+					reg = <0x3>;
+					status = "disabled";
+				};
+
+				switch_port8_base_t1_phy: ethernet-phy@4 {
+					compatible = "ethernet-phy-ieee802.3-c45";
+					reg = <0x4>;
+					status = "disabled";
+				};
+
+				switch_port9_base_t1_phy: ethernet-phy@5 {
+					compatible = "ethernet-phy-ieee802.3-c45";
+					reg = <0x5>;
+					status = "disabled";
+				};
+
+				switch_port10_base_t1_phy: ethernet-phy@6 {
+					compatible = "ethernet-phy-ieee802.3-c45";
+					reg = <0x6>;
+					status = "okay";
+				};
+			};
+
+			mdio@1 {
+				compatible = "nxp,sja1110-base-tx-mdio";
+				#address-cells = <1>;
+				#size-cells = <0>;
+				reg = <1>;
+				status = "okay";
+
+				switch_port1_base_tx_phy: ethernet-phy@1 {
+					reg = <0x1>;
+					status = "okay";
+				};
+
+				switch_port2_base_tx_phy: ethernet-phy@2 {
+					reg = <0x2>;
+					status = "disabled";
+				};
+
+				switch_port3_base_tx_phy: ethernet-phy@3 {
+					reg = <0x3>;
+					status = "disabled";
+				};
+
+				switch_port4_base_tx_phy: ethernet-phy@4 {
+					reg = <0x4>;
+					status = "disabled";
+				};
+			};
+		};
+	};
+};
+
 /* bluetooth */
 &lpuart1 {
 	pinctrl-names = "default";
-- 
2.37.3

